Arch packages implementation

d1nch8g opened the pull request:
close #25037

This PR adds a package registry for [Arch Linux](https://archlinux.org/) packages with support for package files, [signatures](https://wiki.archlinux.org/title/Pacman/Package_signing), and automatic [pacman-database](https://archlinux.org/pacman/repo-add.8.html) management.

Features:

1. Push endpoints to upload package with or without signatures
2. Delete endpoint for specific package version and all related files
3. Get endpoint which is working for package files, gpg signature files and pacman database files
4. Package UI with instructions to connect to the new pacman database and visualised package metadata

![image](https://github.com/go-gitea/gitea/assets/58640116/d2087180-5ceb-4e6b-8162-0532e4ce4bb1)

You can follow [this tutorial](https://wiki.archlinux.org/title/Creating_packages) to build a `*.pkg.tar.zst` package for testing

Co-authored-by: @KN4CK3R 
Co-authored-by: @silverwind

silverwind commented:
The logo is not centered on its viewbox. Suggesting to replace it with https://www.svgrepo.com/svg/349296/arch-linux. I think we do use colored logos for other package types as well.

silverwind commented:
You need to put the svg as `web_src/svg/gitea-arch.svg` and then run `make svg` to get a optimized svg generated that you can use with name `gitea-arch`.

techknowlogick commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,255 @@
+
+// Function that recieves arch package archive data and returns it's metadata.
+func EjectMetadata(filename, domain string, pkg []byte) (*Metadata, error) {
+	d, err := zstd.Decompress(nil, pkg)
```
Thanks for this PR! I haven't gone through it fully yet, but just one thing off the top. Could you use `archiver` for this instead, as I don't believe the datadog import will work on Windows builds.

KN4CK3R commented on routers/init.go:
```diff
@@ -190,6 +190,8 @@ func NormalRoutes() *web.Route {
 		r.Mount("/api/packages", packages_router.CommonRoutes())
 		// This implements the OCI API (Note this is not preceded by /api but is instead /v2)
 		r.Mount("/v2", packages_router.ContainerRoutes())
+		// Arch package routes
+		r.Mount("/api/packages/arch", packages_router.ArchRoutes())
```
This should be in `CommonRoutes` and follow the structure of the other package types.

KN4CK3R commented on services/packages/arch/db_manager.go:
```diff
@@ -0,0 +1,77 @@
+// This function will create new package in database, if it does not exist it
+// will get existing and return it back to user.
+func CreateGetPackage(ctx *context.Context, o *org.Organization, name string) (*pkg.Package, error) {
+	pack, err := pkg.TryInsertPackage(ctx, &pkg.Package{
```
This must use the existing `services/packages` functions.

KN4CK3R commented on services/packages/arch/verificator.go:
```diff
@@ -0,0 +1,110 @@
+	if err != nil {
+		return errors.New("unable to get public keys")
+	}
+	if len(keys) == 0 {
+		return errors.New("no keys for user with email: " + u.Email)
+	}
```
This should not be an error.

The other package types are signed with a generated per-user key. I see no need to verify an upload. If an user is authenticated on the instance it should be enough.

Dancheg97 replied:
I think there should be a way to upload a package with a script or client that uses GnuPG, since arch packages are already signed with gpg, but i can create separate method to make upload using gitea's existing auth system.

I mean it makes sense to use shell script or cli utility when it comes to package uploads, since package delivery is very common operation, and that would be nice if such script/utility would reuse already existing package signature on server side for validation.

Let me know if i should make additional endpoints with standart gitea authorization methods for incoming package validation.

KN4CK3R commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,105 @@
+				<td class="collapsing"><h5>Installed size</h5></td>
+				<td>{{.PackageDescriptor.Metadata.InstalledSizeMib}}</td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Official URL</h5></td>
+				<td> <a target="_blank" rel="noopener noreferrer" href="{{.PackageDescriptor.Metadata.URL}}">{{.PackageDescriptor.Metadata.URL}}</a></td>
+			</tr>
```
This should be in the sidebar.

KN4CK3R commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,105 @@
+				<td class="collapsing"><h5>Description</h5></td>
+				<td>{{.PackageDescriptor.Metadata.Description}}</td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Compressed size</h5></td>
+				<td>{{.PackageDescriptor.Metadata.CompressedSizeMib}}</td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Installed size</h5></td>
+				<td>{{.PackageDescriptor.Metadata.InstalledSizeMib}}</td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Official URL</h5></td>
+				<td> <a target="_blank" rel="noopener noreferrer" href="{{.PackageDescriptor.Metadata.URL}}">{{.PackageDescriptor.Metadata.URL}}</a></td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Build date</h5></td>
+				<td>{{.PackageDescriptor.Metadata.BuildDateStr}}</td>
+			</tr>
+			<tr>
+				<td class="collapsing"><h5>Packager</h5></td>
+				<td>{{.PackageDescriptor.Metadata.Packager}}</td>
+			</tr>
```
This should be in the sidebar.

Dancheg97 replied:
Can i leave URL in the same place? I might not fit into sidebar...

KN4CK3R replied:
Just have a look how it is done for the other package types. The url is not shown, just something like [Project Site](https://gitea.com)

Dancheg97 replied:
![Screenshot from 2023-06-22 11-38-10](https://github.com/go-gitea/gitea/assets/58640116/f2cc613b-f823-4d78-9110-032117bc4f95)
Move URL to side bar, removed build date, compressed and installed size.
Packager does not fit into side bar (it has name and email), would it be ok to leave it on the same place?

KN4CK3R commented on services/packages/arch/file_manager.go:
```diff
@@ -0,0 +1,190 @@
+// or create them if they do not exist.
+func UpdatePacmanDatabases(ctx *context.Context, md *arch.Metadata, distro, owner string) error {
+	// Create temporary directory for arch database operations.
+	tmpdir := path.Join(setting.Repository.Upload.TempPath, uuid.New().String())
+	err := os.MkdirAll(tmpdir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpdir)
```
Why not build that file in memory?

KN4CK3R reviewed (changes requested):
There are some things to address:

1. The routing should follow the current structure of other package types.
2. The storage of blobs must use the existing system.

Dancheg97 commented:
@KN4CK3R 

There is no problem to use the existing system of blob storage for package and signature files, they are not shared across namespace, and belong to the same package. But pacman database files and its symlinks, should be shared across different packages in scope of same organization or user.

May i store pacman database files using unique key made with `distribution`+`architecture`+`database namespace (owner)`+`database file`?

Dancheg97 commented:
I would also like to clarify whether it is possible to make custom authorization for validating incoming packages using package signatures made with GnuPG, since this practice is widespread in the arch community.
As a client for package delivery, I wrote a bash script and a cli utility in go. Bash script and a link to the utility could be added to the documentation.
This does not quite fit with the current authorization methods, but it will be more convenient for the end user, since GnuPG is used for signatures anyway, it may make sense to add authorization in a similar way.

silverwind commented on web_src/svg/octicon-arch.svg:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" viewBox="0 0 128 128" class="svg gitea-arch" width="16" heig ...
```
```suggestion
<svg xmlns="http://www.w3.org/2000/svg" aria-label="Arch Linux" role="img" viewBox="0 0 512 512"><rect width="512" height="512" rx="15%"/><path d="M256 72c-14 35-23 57-39 91 10 11 22 23 41 36-21-8-35-17-45-26-21 43-53 103-117 220 50-30 90-48 127-55-2-7-3-14-3-22v-1c1-33 18-58 38-56 20 1 36 29 35 62l-2 17c36 7 75 26 125 54l-27-50c-13-10-27-23-55-38 19 5 33 11 44 17-86-159-93-180-122-250z" fill="#1793d1"/></svg>
```

Then run `make svg` afterwards.

KN4CK3R commented on go.mod:
```diff
@@ -16,6 +16,7 @@ require (
 	github.com/42wim/sshsig v0.0.0-20211121163825-841cf5bbc121
 	github.com/Azure/go-ntlmssp v0.0.0-20221128193559-754e69321358
 	github.com/NYTimes/gziphandler v1.1.1
+	github.com/ProtonMail/gopenpgp/v2 v2.7.1
```
We already have and use `github.com/keybase/go-crypto/openpgp`.

KN4CK3R commented on modules/packages/content_store.go:
```diff
@@ -63,3 +64,17 @@ func RelativePathToKey(relativePath string) (BlobHash256Key, error) {
 
 	return BlobHash256Key(parts[2]), nil
 }
+
+// Save data with specified string key.
+func (s *ContentStore) SaveStrBytes(key string, data []byte) error {
+	return s.Save(BlobHash256Key(key), bytes.NewReader(data), int64(len(data)))
+}
+
+// Get data related to provided key.
+func (s *ContentStore) GetStrBytes(key string) ([]byte, error) {
+	obj, err := s.Get(BlobHash256Key(key))
+	if err != nil {
+		return nil, err
+	}
+	return io.ReadAll(obj)
+}
```
`GetStrBytes` is used but `SaveStrBytes` is not, so `GetStrBytes` will always fail. Additionally both methods should not be needed.

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,330 @@
+			return nil, err
+		}
+	}
+	splt := strings.Split(buf.String(), "PKGINFO")
+	if len(splt) < 2 {
+		return nil, errors.New("unable to eject .PKGINFO from archive")
+	}
+	raw := splt[1][0:10000]
```
This looks really fragile. The package is just a compressed tar archive, so unpack it and process the `PKINFO` file without reinpreting everything as a string.

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,330 @@
+	}
+	raw := splt[1][0:10000]
+	inssize := int64(len(pkg))
+	compsize := ejectInt64(raw, "size")
+	unixbuilddate := ejectInt64(raw, "builddate")
+	return &Metadata{
+		Filename:          filename,
+		Name:              ejectString(raw, "pkgname"),
+		Base:              ejectString(raw, "pkgbase"),
+		Version:           ejectString(raw, "pkgver"),
+		Description:       ejectString(raw, "pkgdesc"),
+		CompressedSize:    inssize,
+		CompressedSizeMib: ByteCountSI(inssize),
+		InstalledSize:     compsize,
+		InstalledSizeMib:  ByteCountSI(compsize),
+		MD5:               md5sum(pkg),
+		SHA256:            sha256sum(pkg),
+		URL:               ejectString(raw, "url"),
+		BuildDate:         unixbuilddate,
+		BuildDateStr:      ReadableTime(unixbuilddate),
+		BaseDomain:        domain,
+		Packager:          ejectString(raw, "packager"),
+		Provides:          ejectStrings(raw, "provides"),
+		License:           ejectStrings(raw, "license"),
+		Arch:              ejectStrings(raw, "arch"),
+		Depends:           ejectStrings(raw, "depend"),
+		OptDepends:        ejectStrings(raw, "optdepend"),
+		MakeDepends:       ejectStrings(raw, "makedepend"),
+		CheckDepends:      ejectStrings(raw, "checkdepend"),
+		Backup:            ejectStrings(raw, "backup"),
```
The `eject` functions should just be a switch for the key inside a `Scanner` loop to read the input line by line.

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,330 @@
+	return i
+}
+
+func ByteCountSI(b int64) string {
+	const unit = 1000
+	if b < unit {
+		return fmt.Sprintf("%d B", b)
+	}
+	div, exp := int64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB", float64(b)/float64(div), "kMGTPE"[exp])
+}
```
The fields which use this function are unused. If there must be a human readable display of the size, the conversion should be done in UI.

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,330 @@
+	return fmt.Sprintf("%.1f %cB", float64(b)/float64(div), "kMGTPE"[exp])
+}
+
+func ReadableTime(unix int64) string {
+	return time.Unix(unix, 0).Format(time.DateTime)
+}
```
Same here.

Dancheg97 commented:
@KN4CK3R 

Changed implementation. resulting PR should contain:

Dancheg97 commented:
@KN4CK3R  @techknowlogick 

Created implementation which includes:
1) Endpoint to push arch packages
2) Endpoint to get arch packages
3) Endpoint to get userspace relevant pacman database
4) Endpoint to delete packages
5) Documentation for each endpoint and database connection mechanics
6) Tests for push/package get/database get/remove operations with scenario with single package with different architectures (in tests used library to mock time, bou.ke/monkey)
7) Gitea user/org space with arch packages can be automatically connected in /etc/pacman.conf
8) Pacman database is created on demand, cached for 5 minutes.

Push and delete operations are validated with user's GnuPG keys.

Is there anything else i should cover in this PR?

silverwind commented:
Waiting on @KN4CK3R to review.

Dancheg97 commented:
@artizirk
I thought it would be a good idea to force people to use `gpg`, since it is a good practice...
Will make a draft with existing auth methods.

KN4CK3R commented:
Only Arch uses them ;-) Python for example removed it years ago.

Dancheg97 commented:
@artizirk @KN4CK3R 
> Can I use the same ssh key HTTP Signatures for package registry?

Should be possible, `OAuth2` exists in authorization methods for packages endpoints. (`routers/api/packages/api.go`)

```go
// CommonRoutes provide endpoints for most package managers (except containers - see below)
// These are mounted on `/api/packages` (not `/api/v1/packages`)
func CommonRoutes() *web.Route {
	r := web.NewRoute()

	r.Use(context.PackageContexter())

	verifyAuth(r, []auth.Method{
		&auth.OAuth2{},
		&auth.Basic{},
		&nuget.Auth{},
		&conan.Auth{},
		&chef.Auth{},
	})
```

> Can the HTTP Signatures spec be extended to also include OpenPGP signatures support so that all of Gitea's API's could be used with it?

Shouldn't be impossible, need to learn more about [OAuth2](https://github.com/oauth-wg/oauth-v2-1) with [http signatures](https://oauth.net/http-signatures/). At the moment seems like it is not supported.

> Is Gitea the only one who has implemented HTTP Signatures?

No for sure :D

> This pull-request probably shouldn't invent its own PGP based auth and just use the existing Gitea auth methods.

Should have better learned existing authentication methods. Didn't know about that trick with `gpg-agent`, that is really cool, need to play around with it.

---

Will rework arch package endpoints to use existing `gpg-agent` authorization from OAuth2

Can i also create PR in [tea](https://gitea.com/gitea/tea) with with command for generic/arch package push method?

Dancheg97 commented:
@KN4CK3R

- Reworked API to use current authentication methods
- Corrected tests, arch package documentation

artizirk commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,111 @@
+			<div class="field">
+				<label>{{svg "octicon-note"}} {{.locale.Tr "packages.arch.pacmanconf"}}</label>
+				<div class="markup"><pre class="code-block"><code>[{{.PackageDescriptor.Owner.LowerName}}.{{.PackageDescriptor.Metadata.BaseDomain}}]
+Server = https://{{.PackageDescriptor.Metadata.BaseDomain}}/api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64</code></pre></div>
```
This should probably use `{{AppSubUrl}}` Like in here https://github.com/go-gitea/gitea/blob/a12a5f3652c339b17b187ff424a480631a3c1e1e/templates/package/content/debian.tmpl#L7

Dancheg97 replied:
Yes, i was trying that before, but the result was empty URL. Do i have to add some additional parameters/apply specific condition for `{{AppSubUrl}}` variable to work properly?
![image](https://github.com/go-gitea/gitea/assets/58640116/1a1f6ed5-a114-46fc-9464-ffb6456887fd)

artizirk replied:
Looks like it has to be part of [`gitea-origin-url`](https://github.com/go-gitea/gitea/blob/cdd3d4b8d8814137e16116002e5bf6aad38b794f/web_src/js/webcomponents/GiteaOriginUrl.js#L4) custom html element

```
diff --git a/templates/package/content/arch.tmpl b/templates/package/content/arch.tmpl
index 5f5c8cd49791..8d82265056bf 100644
--- a/templates/package/content/arch.tmpl
+++ b/templates/package/content/arch.tmpl
@@ -6,7 +6,7 @@
                        <div class="field">
                                <label>{{svg "octicon-note"}} {{.locale.Tr "packages.arch.pacmanconf"}}</label>
                                <div class="markup"><pre class="code-block"><code>[{{.PackageDescriptor.Owner.LowerName}}.{{.PackageDescriptor.Metadata.BaseDomain}}]
-Server = https://{{.PackageDescriptor.Metadata.BaseDomain}}/api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64</code></pre></div>
+Server = <gitea-origin-url data-url="{{AppSubUrl}}/api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64"></gitea-origin-url></code></pre></div>
                        </div>

                        <div class="field">
```

artizirk replied:
The reason why there is a custom element is explained in the pull request https://github.com/go-gitea/gitea/pull/22861
TLDR: it's hard for server side to always know the correct user visible url so its calculated in the browser.

Dancheg97 replied:
@artizirk 
Corrected `arch.tmpl`, to use `{{AppURL}}` variable:
```html
<div class="markup"><pre class="code-block"><code>[{{.PackageDescriptor.Owner.LowerName}}.{{.PackageDescriptor.Metadata.BaseDomain}}]
SigLevel = Optional TrustAll
Server = {{AppUrl}}api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64</code></pre></div>
</div>
```
I would also like to remove `{{.PackageDescriptor.Metadata.BaseDomain}}` variable from template. Is there any way to get domain name in UI?

Dancheg97 replied:
Changed template to use `{{.RegistryHost}}` variable, similarly with container registry.

```html
<div class="markup"><pre class="code-block"><code>[{{.PackageDescriptor.Owner.LowerName}}.{{.RegistryHost}}]
SigLevel = Optional TrustAll
Server = {{AppUrl}}api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64</code></pre></div>
</div>
```

artizirk commented:
I tested this pull request on my machine and it works! Thank you @Dancheg97 ❤️

Only change I had to do was to strip newlines from the package signature xxd command output as curl or gitea web server did not like multi line http headers. Full curl command that worked for me looks like this.
```
curl -v -X PUT http://arti:testpass@localhost:3000/api/packages/arti/arch/push \
 --data-binary '@swaynag-battery-0.1.1-1-x86_64.pkg.tar.zst' \
 --header 'Content-Type: application/octet-stream' \
 --header 'distro: archlinux' \
 --header 'filename: swaynag-battery-0.1.1-1-x86_64.pkg.tar.zst' \
 --header "sign: $(xxd -p swaynag-battery-0.1.1-1-x86_64.pkg.tar.zst.sig | tr -d '\n')"
```

Dancheg97 commented:
@artizirk @KN4CK3R 

Also, I wanted to highlight a function which I use in the `push` method. This function automatically connects the repository with the source code to the package after the blob upload if there is a repository with the same name in gitea for instance.

```go
// Automatically connect repository for provided package if name matched.
err = arch_service.RepositoryAutoconnect(ctx, owner, md.Name, pkgid)
if err != nil {
	apiError(ctx, http.StatusInternalServerError, err)
	return
}
```

Maybe it would be better to create a unified approach for that scenario.

Dancheg97 commented:
@artizirk

I am also not sure about `{{AppUrl}}` and `{{AppSubUrl}}` variables. `{{AppUrl}}` is working locally, while `{{AppSubUrl}}` is working on my public instance. Why are they different, and which is better for package templates?

delvh commented:
```go
// AppURL is the Application ROOT_URL. It always has a '/' suffix
// It maps to ini:"ROOT_URL"
AppURL string
// AppSubURL represents the sub-url mounting point for gitea. It is either "" or starts with '/' and ends without '/', such as '/{subpath}'.
// This value is empty if site does not have sub-url.
AppSubURL string
```

Dancheg97 commented:
@delvh 

If i use `{{AppSubUrl}}` i get following result:

(local instance)
![Screenshot from 2023-07-25 22-09-23](https://github.com/go-gitea/gitea/assets/58640116/f399d824-e5ff-4b89-899d-ed001589fe24)
(public instance)
![Screenshot from 2023-07-25 22-19-17](https://github.com/go-gitea/gitea/assets/58640116/196bbb91-a624-4efa-87a9-4fa6e0ffcf14)


With `{{AppURL}}` i can see full link:

(local instance)
![Screenshot from 2023-07-25 22-14-05](https://github.com/go-gitea/gitea/assets/58640116/46b6671f-b7c4-4500-9c39-a3070eceeb17)
(public instance)
![image](https://github.com/go-gitea/gitea/assets/58640116/8077e738-5a6c-49c4-b96d-5e5b5f12290c)

delvh commented:
Yep, that's exactly one of the few cases where you need `AppUrl`.
Those cases are the reason why it isn't removed yet.
You need a full URL to clone from, to direct API calls to, or to download anything else from.
Packages fit this criteria.

lunny commented:
You can also use the domain users requested in the web browser to support multiple domains.

KN4CK3R commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,112 @@
+			</div>
+
+			<div class="field">
+				<label>{{.locale.Tr "packages.arch.documentation" "https://wiki.archlinux.org/title/mirrors" | Safe}}</label>
```
Should link to the Gitea docs.

KN4CK3R commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,112 @@
+			</tr>
+			{{end}}
+
+			{{range $key := .PackageDescriptor.Metadata.Depends}}
+			<tr>
+				<td class="collapsing"><h5>Depends</h5></td>
+				<td>{{$key}}</td>
+			</tr>
+			{{end}}
```
I don't think we need the "header" every time.

KN4CK3R commented on tests/integration/api_packages_arch_test.go:
```diff
@@ -0,0 +1,472 @@
+		req.Header.Set("distro", "archlinux")
+		req.Header.Set("pkgsign", hex.EncodeToString(firstPackageSignatureData))
+		req.Header.Set("Content-Type", "application/octet-stream")
+		req.Body = io.NopCloser(bytes.NewReader(firstPackageData))
```
Just use `NewRequestWithBody`.

KN4CK3R commented on tests/integration/api_packages_arch_test.go:
```diff
@@ -0,0 +1,472 @@
+
+	assert.NoError(t, unittest.PrepareTestDatabase())
+
+	setting.Packages.LimitSizeArch = 99999999999999
```
```suggestion
```
Default is not limited.

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,160 @@
+		return
+	}
+
+	ctx.Resp.WriteHeader(http.StatusOK)
```
```suggestion
	ctx.Status(http.StatusOK)
```

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,160 @@
+// Remove specific package version, related files and pacman database entry.
+func Remove(ctx *context.Context) {
+	var (
+		pkg = ctx.Req.Header.Get("package")
+		ver = ctx.Req.Header.Get("version")
```
Just use url parameters then you can use the methods the framework provides. Adding headers to requests is tedious.

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,160 @@
+		return
+	}
+
+	ctx.Resp.WriteHeader(http.StatusNotFound)
```
```suggestion
	ctx.Status(http.StatusNotFound)
```

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,160 @@
+
+		ctx.ServeContent(pkg, &context.ServeHeaderOptions{
+			Filename:      file,
+			CacheDuration: time.Minute * 5,
```
```suggestion
```

KN4CK3R commented on services/packages/arch/db_manager.go:
```diff
@@ -0,0 +1,210 @@
+// This function creates new package, version and package_file properties in
+// database, and writes blob to file storage. If package/version/blob exists it
+// will overwrite existing data. Package id and error will be returned.
+func SaveFile(ctx *context.Context, p *SaveFileParams) (int64, error) {
```
I see no benefit of this helper method.

Dancheg97 replied:
@KN4CK3R 
I save 2 files (3 if signature provided) for each package. Which technique should i use to reduce code repetition?

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,160 @@
+	)
+
+	// Read package to memory for package validation.
+	pkgdata, err := io.ReadAll(ctx.Req.Body)
```
It's not save to read the full file into memory. Use `CreateHashedBufferFromReader`.

KN4CK3R commented on modules/setting/packages.go:
```diff
@@ -24,6 +24,7 @@ var (
 		LimitTotalOwnerCount int64
 		LimitTotalOwnerSize  int64
 		LimitSizeAlpine      int64
+		LimitSizeArch        int64
```
You are missing the set-ting later in the file.

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,280 @@
+// This function creates a list containing unique values formed of 2 passed
+// slices.
+func UnifiedList(first, second []string) []string {
+	unique := map[string]struct{}{}
+	for _, v := range first {
+		unique[v] = struct{}{}
+	}
+	for _, v := range second {
+		unique[v] = struct{}{}
+	}
+	var archs []string
+	for k := range unique {
+		archs = append(archs, k)
+	}
+	return archs
```
```suggestion
	unique := make(Set[string], len(first)+len(second))
	s.AddMultiple(first...)
	s.AddMultiple(second...)
	return s.Values()
```

delvh replied:
`s`? Or `unique`?

delvh replied:
```suggestion
	set := make(Set[string], len(first)+len(second))
	set.AddMultiple(first...)
	set.AddMultiple(second...)
	return set.Values()
```

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,280 @@
+// Join database or package names to prevent collisions with same packages in
+// different user spaces. Skips empty strings and returns name joined with
+// dots.
+func Join(s ...string) string {
```
Not used?

KN4CK3R commented on docs/content/doc/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,68 @@
+---
+date: "2016-11-08T16:00:00+02:00"
+title: "Title"
+weight: 10
+toc: true
+draft: false
+menu:
+  sidebar:
+    parent: "packages"
+    name: "Arch"
+    weight: 10
+    identifier: "arch"
```
Wrong title and must use new docusaurus structure.

Dancheg97 replied:
@KN4CK3R 
Moved file to new packages folder, fixed title. Are there any other changes to the documentation needed?

Dancheg97 commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,280 @@
+	return ``
+}
+
+func rmEmptyStrings(s []string) []string {
```
used in `GetDbDesc()`
```go
// This function returns pacman package description in unarchived raw database
// format.
func (m *Metadata) GetDbDesc() string {
	return strings.Join(rmEmptyStrings([]string{
		formatField("FILENAME", m.Filename),
		formatField("NAME", m.Name),
		formatField("BASE", m.Base),
		formatField("VERSION", m.Version),
		formatField("DESC", m.Description),
		formatField("CSIZE", m.CompressedSize),
		formatField("ISIZE", m.InstalledSize),
		formatField("MD5SUM", m.MD5),
		formatField("SHA256SUM", m.SHA256),
		formatField("URL", m.URL),
		formatField("LICENSE", m.License),
		formatField("ARCH", m.Arch),
		formatField("BUILDDATE", m.BuildDate),
		formatField("PACKAGER", m.Packager),
		formatField("PROVIDES", m.Provides),
		formatField("DEPENDS", m.Depends),
		formatField("OPTDEPENDS", m.OptDepends),
		formatField("MAKEDEPENDS", m.MakeDepends),
		formatField("CHECKDEPENDS", m.CheckDepends),
	}), "\n\n") + "\n\n"
}
```

Dancheg97 commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,112 @@
+				<label>{{svg "octicon-note"}} {{.locale.Tr "packages.arch.pacmanconf"}}</label>
+				<div class="markup"><pre class="code-block"><code>[{{.PackageDescriptor.Owner.LowerName}}.{{.RegistryHost}}]
+SigLevel = Optional TrustAll
+Server = {{AppUrl}}api/packages/{{.PackageDescriptor.Owner.LowerName}}/arch/archlinux/x86_64</code></pre></div>
```
@KN4CK3R
I'm not sure about the best way to do this for the large number of possible combinations of architectures and distributions. If a certain package supports 2 distributions and 2 architectures, there should be 4 different links (and `any` architecture should also be treated as unique case, links for packages with `any` architecture should be generated for all possible architectures). Archlinux and x86_64 are the default options, so I set this option. Also it would be much easier to handle those parameters on client side (since client knows it's architecture and distribution).

artizirk replied:
Or you could just leave it like this
```
Server = <gitea-origin-url data-url="{{AppSubUrl}}/api/packages/{{.PackageDescriptor.Owner.Name}}/arch/$distribution/$arch"></gitea-origin-url></code></pre></div>
```

and add instructions to replace `$distribution` with value that can be taken from `.PackageDescriptor.Metadata.DistroArch` table row from the bottom of the page.

`$arch` gets replaced by `pacman` with current architecture as is documented in the [pacman.conf(5) man page](https://man.archlinux.org/man/pacman.conf.5.en#REPOSITORY_SECTIONS)

KN4CK3R replied:
Then just let the user decide. Debian example:
![grafik](https://github.com/go-gitea/gitea/assets/1666336/8551782f-d418-444f-a9e2-4dc2dad3edf9)

The list below is just space seperated if there is more than one value. If not every version of a package is available in every possible combination thats bad luck...

Dancheg97 commented:
@KN4CK3R 
I've made small refactoring for `reatePackageAndAddFile` and `createPackageAndAddFile` functions `services/packages/packages.go:83` (they are the same). I need `PackageBlob` as returned value (to get compressed size, sha256sum and md5sum) for pacman database.

KN4CK3R commented:
Please revert that. You can already access the hashes from the `buf` you created with
`buf, err := pkg_module.CreateHashedBufferFromReader(tee)`

```go
hashMD5, hashSHA1, hashSHA256, hashSHA512 := buf.Sums()
_ = hex.EncodeToString(hashMD5),
_ = hex.EncodeToString(hashSHA1),
_ = hex.EncodeToString(hashSHA256),
_ = hex.EncodeToString(hashSHA512),
```

```go
var ejectbuf bytes.Buffer
tee := io.TeeReader(upload, &ejectbuf)
```
still reads all the incoming data into memory. You can simply pass `buf` as reader into `EjectMetadata` and rewind the reader afterwards with `buf.Seek(0, io.SeekStart)`.

Dancheg97 commented:
@KN4CK3R 

When I try to upload 1 package (containing 3 files) multiple times, there are multiple entries in the database corresponding to each package file. This behavior is not entirely clear to me, because the hash sums of identical files do not match. I also don't understand why deleting a package doesn't remove the file/hash sum records from the database. Link to [package](https://fmnx.su/dancheg97/package), upload script:

```sh
makepkg -sfr --sign --noconfirm
curl -X PUT \
  "http://localhost:3000/api/packages/dancheg97/arch/push/package-1-1-x86_64.pkg.tar.zst/archlinux/$(xxd -p package-1-1-x86_64.pkg.tar.zst.sig | tr -d '\n')" \
  --header "Content-Type: application/octet-stream" \
  --header "dancheg97:gitgit" \
  --data-binary '@package-1-1-x86_64.pkg.tar.zst'
```

![image](https://github.com/go-gitea/gitea/assets/58640116/0baee53b-bf25-4085-8f94-59a2237775fa)

I tried to delete the package with the button in ui:

![image](https://github.com/go-gitea/gitea/assets/58640116/679f3e70-47ac-4fb3-a133-f7c716a63eb9)

KN4CK3R commented:
> When I try to upload 1 package (containing 3 files) multiple times, there are multiple entries in the database corresponding to each package file. This behavior is not entirely clear to me, because the hash sums of identical files do not match in database.

I don't understand what you did. If you upload the same file twice into the same arch/dist combination this should be restricted. If files have different names there should not be an error. `n` identical files with different names result in `n` `package_file` entries and one `package_blob` entry.

> I also don't understand why deleting a package doesn't remove the blob records from the database.

See my comment here: https://github.com/go-gitea/gitea/issues/25035#issuecomment-1589312885

ChewingBever commented:
I've got some questions about this PR

* The screenshots seem to suggest that multiple versions of a package can be published, but how does this work considering Pacman repositories are only allowed to have one version of any given package?
* What routes are the `.db` and `.files` served on? I've tried locating the one for your `core` organisation on your, but failed to find the proper route
* Does this PR create a single repository per user/organisation or one for each package?

Dancheg97 commented:
@ChewingBever 

> The screenshots seem to suggest that multiple versions of a package can be published, but how does this work considering Pacman repositories are only allowed to have one version of any given package?

Multiple versions of a package can be published, when database is requested - it is created with the following actions: each version starting from is `latest` to the `oldest` is checked to be compatible with the requested combination of architecture and distribution. When the first compatible package is found, the related `desc` file will be added to the database archive and 'search' stops. As a result, only a single `desc` file related to each package will be in the database archive, and it will be compatible with the requested combination of distribution and architecture.

Example: [database download link](https://fmnx.su/api/packages/core/arch/archlinux/x86_64/core.fmnx.su.db)

> What routes are the .db and .files served on? I've tried locating one for your core organisation on your website, but failed to find the proper route.

They are served on the same routes with package files, `/api/packages/{owner}/arch/{distribution}/{architecture}/{filename}`. This can be used as a link in your `pacman.conf` file.

> Does this PR create a single repository per user/organisation or one for each package?

Repositories are created for each use/organization. (bottom link contains multiple packages from the `core` organization of my public gita for instance)

Dancheg97 commented:
@KN4CK3R @artizirk 

Made some changes and fixes in PR:

1) Distro/arch specific parameters are no longer saved in JSON metadata in the SQL database, package `desc` files are created on package upload and saved in object storage with other package files. When creating a pacman database, those files are used.

![Screenshot from 2023-08-03 14-43-14](https://github.com/go-gitea/gitea/assets/58640116/205c1caf-3733-40b2-a9a6-44a7258b15c6)

2) Reduced size of package files in tests, added tests to check package signatures. Created [repository](https://fmnx.su/dancheg97/package) which contains a script to create test data inside a docker container. Should I save this data in gitea repo (in case some changes are needed)?

3) All parameters moved from headers to URL params.

4) Documentation corrections - arch packages in overview, path URL params in links. I am not sure if this line is correct in documentation, how. How was it created?

```md
---
date: "2016-11-08T16:00:00+02:00"
Title: "Arch Package Registry"
weight: 10
toc: true
draft: false
menu:
  sidebar:
    parent: "packages"
    Name: "Arch"
    weight: 10           <<< <<< here
    identifier: "arch"
---
```

5) I save 3 files for each package (package, signature and pacman desc file). Should I unwrap the `SaveFile` function or unify this logic in a different way?

```go
// Creates new package, version and package_file properties in database,
// and writes blob to file storage. If package/version/blob exists it
// will overwrite existing data. Package id and error will be returned.
func SaveFile(ctx *context.Context, p *SaveFileParams) (int64, error) {
	ver, _, err := pkg_service.CreatePackageOrAddFileToExisting(
		&pkg_service.PackageCreationInfo{
			PackageInfo: pkg_service.PackageInfo{
				Owner:       p.Owner,
				PackageType: pkg_model.TypeArch,
				Name:        p.PkgName,
				Version:     p.PkgVer,
			},
			Creator:  p.Creator,
			Metadata: p.Metadata,
		},
		&pkg_service.PackageFileCreationInfo{
			PackageFileInfo: pkg_service.PackageFileInfo{
				Filename:     p.Filename,
				CompositeKey: p.Distro + "-" + p.Filename,
			},
			Creator:           p.Creator,
			Data:              p.Buf,
			OverwriteExisting: true,
			IsLead:            p.IsLead,
		},
	)
	if err != nil {
		return 0, err
	}
	return ver.ID, nil
}
```

Dancheg97 commented:
@ChewingBever @artizirk 

> Vieter is my personal solution to a problem I’ve been facing for months: extremely long AUR package build times. 

I've read the documentation for your project [https://github.com/vieter-v/vieter]. I was trying to solve the same problem with nearly the same approach ([first attempt](https://github.com/ctlos/ctlpkg), [second attempt](https://fmnx.su/dancheg97/repo)).
It would be nice to have a generic API for the Arch packages. Gitea seems much better because you can store the project source code, documentation and related package builds in the same place.

Also, I've made [CLI utility](https://fmnx.su/core/pack) that can build AUR packages (or any git repo) and push them to gitea:

```sh
pack -Bqs aur.archlinux.org/nvm aur.archlinux.org/etcher-bin
pack -P fmnx.su/core/nvm fmnx.su/core/etcher-bin
```

I use this tool to update packages in pacman registry for ISO image.

KN4CK3R commented:
Haven't looked into it any further but you don't need that desc file. You could just store in a property like I do with Debian packages:
https://github.com/go-gitea/gitea/blob/952882fe4ab6950ca0fb33af3885cb3347d46e78/routers/api/packages/debian/debian.go#L180-L184

ChewingBever commented:
@Dancheg97 

thanks for looking at my project! yeah i wanted vieter (or rather, rieter in the future) to be a standalone implementation of a repository server, as i wanted it to be used to host full distros one day (perhaps my own mirrors of stuff), but i think it would definitely pair well with gitea well for hosting the PKGBUILD repositories (although perhaps for simply those, i would prefer something even more lightweight like cgit).

Dancheg97 commented:
> as i wanted it to be used to host full distros one day

It seems to me that it is also possible to adapt the gitea for these purposes. Now users are forced to switch between different services to work on the codebase / provide documentation / provide package builds. Gitea can be used for all these purposes.

Standalone implementation is fine too, but merging these concepts into a CI-CD system seems more convenient, as fewer systems need to be used to provide programs (and related documentation) to users.

Dancheg97 commented:
@KN4CK3R 

Changed implementation to use package file properties.

Dancheg97 commented:
@KN4CK3R

Changed the way package signatures and database metadata files are created and stored for arch packages.

Each time new arch package is uploaded - new `.desc` file for pacman database is created (from ejected package metadata) and stored in package properties alongside with signature hex.

```go
	props := map[string]string{
		distro + "-" + filename + ".desc": desc.String(),
	}
	if sign != "" {
		_, err := hex.DecodeString(sign)
		if err != nil {
			apiError(ctx, http.StatusBadRequest, err)
		}
		props[distro+"-"+filename+".sig"] = sign
	}
```

There are a few things which i would like to clarify, since i am not sure about which solution is better:

1) Endpoint to get arch package files, contains 3 options: get package files, get package signatures and get package database. With package files `.pkg.tar.zst` and signatures `.pkg.tar.zst.sig` it is simple: function just makes lookup to object storage or package file properties. But package database `.db` or `.db.tar.gz` is created on demand. Current implementation makes many SQL queries to get `.desc` properties for all packages in specific organization/user scope. The question is whether it should be optimized to single SQL query?

2) Currently i have not implemented any mechanism to receive database signatures. This could be made if there would be an option to store private GPG key with password, related to specific organization in gitea, and sign database files 'on-fly'. (Similarly, as github signs commits with github GPG verified signature) Should this feature be implemented/ added in same PR?

3) Currently, there is a single endpoint to push arch packages:

```go
		r.Group("/arch", func() {
			r.Put("/push/{filename}/{distro}/{sign}", arch.Push, reqPackageAccess(perm.AccessModeWrite)) // << here
			r.Delete("/remove/{package}/{version}", arch.Remove, reqPackageAccess(perm.AccessModeWrite))
			r.Get("/{distro}/{arch}/{file}", arch.Get)
		})
```

And there is a check for package signature (if present in request, it would be saved in package properties as hex string).

```go
	props := map[string]string{
		distro + "-" + filename + ".desc": desc.String(),
	}
	if sign != "" {
		_, err := hex.DecodeString(sign)
		if err != nil {
			apiError(ctx, http.StatusBadRequest, err)
		}
		props[distro+"-"+filename+".sig"] = sign
	}
```

Should i have different endpoint for scenario where package signature is not pused?

```go
		r.Group("/arch", func() {
			r.Put("/push/{filename}/{distro}", arch.Push, reqPackageAccess(perm.AccessModeWrite)) // << without signature
			r.Put("/push/{filename}/{distro}/{sign}", arch.Push, reqPackageAccess(perm.AccessModeWrite))
			r.Delete("/remove/{package}/{version}", arch.Remove, reqPackageAccess(perm.AccessModeWrite))
			r.Get("/{distro}/{arch}/{file}", arch.Get)
		})
```


4) Should PR have an option for private arch packages? I am not sure if there is a need in such feature because there is still an option host private gitea instance for such purpose.

Dancheg97 commented:
@artizirk @ChewingBever @KN4CK3R @techknowlogick 

To sum up, currently there are three endpoints:

1) Package upload: creates a package with related properties and parameters required for visual representation. Also saves the database `.desc` file and package `.pkg.tar.zst.sig` file in properties attached to the file.
2) Package download: single endpoints that will handle package files, package signatures, and Pacman database files. Packages are downloaded from object storage, signatures from package properties, and pacman database files are created in user or organization scope from all existing arch packages in accordance with the requested combination of architecture and distribution.
3) Package remove: a simple endpoint to remove the package version and all it's files.

I wanted to ask if other endpoints should be added or if current endpoints should be reworked in a different manner?

KN4CK3R commented on models/packages/package_file.go:
```diff
@@ -25,6 +25,8 @@ var (
 	ErrDuplicatePackageFile = util.NewAlreadyExistErrorf("package file already exists")
 	// ErrPackageFileNotExist indicates a package file not exist error
 	ErrPackageFileNotExist = util.NewNotExistErrorf("package file does not exist")
+	// ErrPackagePropertyNotExist indicates a package property not exist error
+	ErrPackagePropertyNotExist = util.NewNotExistErrorf("package property does not exist")
```
not used
```suggestion
```

d1nch8g replied:
Removed

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,248 @@
+		}
+	}
+
+	tw.Close()
+	gw.Close()
```
Duplicate close with the defers.

KN4CK3R commented on routers/api/packages/api.go:
```diff
@@ -122,6 +123,11 @@ func CommonRoutes() *web.Route {
 			})
 		}, reqPackageAccess(perm.AccessModeRead))
+		r.Group("/arch", func() {
+			r.Put("/push/{filename}/{distro}/{sign}", arch.Push, reqPackageAccess(perm.AccessModeWrite))
+			r.Delete("/remove/{package}/{version}", arch.Remove, reqPackageAccess(perm.AccessModeWrite))
```
The permission checks must be first.

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	if strings.HasSuffix(file, ".db.tar.gz") || strings.HasSuffix(file, ".db") {
+		db, err := arch_service.CreatePacmanDb(ctx, &arch_service.DbParams{
+			Owner:        owner,
+			Architecture: arch,
+			Distribution: distro,
+		})
+		if err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+			return
+		}
+
+		ctx.ServeContent(bytes.NewReader(db), &context.ServeHeaderOptions{
```
`CreatePacmandDb` uses a buffer internally but returns `[]byte` just to create a new reader. Could just return the buffer or `io.Reader`.

Dancheg97 replied:
@KN4CK3R 

There is no `Seek()` method to pass `bytes.Buffer` directly to `ctx.ServeContent()`. Replaced with `HashedBuffer` from packages `module`.

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,204 @@
+		&pkg_service.PackageFileCreationInfo{
+			PackageFileInfo: pkg_service.PackageFileInfo{
+				Filename:     filename,
+				CompositeKey: distro + "-" + filename,
```
Adding the filename does not change anything or?

Dancheg97 replied:
To ensure that composite key is unique for different distributions for the same file name.

Dancheg97 replied:
Fixed, removed unnecessary filename from composite key.

KN4CK3R commented on services/packages/arch/service.go:
```diff
@@ -0,0 +1,182 @@
+
+// Get data related to provided filename and distribution, for package files
+// update download counter.
+func GetPackageFile(ctx *context.Context, distro, file string) (storage.Object, error) {
+	pkgfile := &pkg_model.PackageFile{CompositeKey: distro + "-" + file}
+
+	ok, err := db.GetEngine(ctx).Get(pkgfile)
+	if err != nil || !ok {
+		return nil, fmt.Errorf("%+v %t", err, ok)
+	}
+
+	b, err := pkg_model.GetBlobByID(ctx, pkgfile.BlobID)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasSuffix(file, ".pkg.tar.zst") {
+		err = pkg_model.IncrementDownloadCounter(ctx, pkgfile.VersionID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return packages.NewContentStore().Get(packages.BlobHash256Key(b.HashSHA256))
+}
```
Why is there no version parameter? This method should use `package_service.GetPackageFileStream` or be replaced with one of the other helper methods.

Dancheg97 replied:
@KN4CK3R 

Arch package files contain version, release number and architecture in file name. For example:

```
zsh-5.9-4-x86_64.pkg.tar.zst
zsh-autosuggestions-0.7.0-2-any.pkg.tar.zst
zsh-syntax-highlighting-0.7.1-1-any.pkg.tar.zst
```

I have 2 options to get package file property from database:

- Using existing helper methods:
```go
func GetPackageFile(ctx *context.Context, distro, file string) (io.ReadSeekCloser, error) {
	var (
		splt     = strings.Split(file, "-")
		pkgname  = strings.Join(splt[0:len(splt)-3], "-")
		vername  = splt[len(splt)-3] + "-" + splt[len(splt)-2]
		filename = strings.TrimSuffix(file, ".sig")
		filekey  = distro + "-" + filename
	)

	version, err := pkg_model.GetVersionByNameAndVersion(
		ctx, ctx.Package.Owner.ID, pkg_model.TypeArch, pkgname, vername,
	)
	if err != nil {
		return nil, err
	}

	pkgfile, err := pkg_model.GetFileForVersionByName(
		ctx, version.ID, filename, filekey,
	)
	if err != nil {
		return nil, err
	}

	filestream, _, _, err := pkg_service.GetPackageFileStream(ctx, pkgfile)
	return filestream, err
}
```

- Using direct call to database with package composite key:
```go
func GetPackageFile(ctx *context.Context, distro, file string) (io.ReadSeekCloser, error) {
	pkgfile := &pkg_model.PackageFile{CompositeKey: distro + "-" + file}

	ok, err := db.GetEngine(ctx).Get(pkgfile)
	if err != nil || !ok {
		return nil, fmt.Errorf("%+v %t", err, ok)
	}

	filestream, _, _, err := pkg_service.GetPackageFileStream(ctx, pkgfile)
	return filestream, err
}
```

Which one would be better for current scenario?

d1nch8g replied:
Added function that gets package file property from file name. It is possible, since file name contains all required information.

Dancheg97 commented:
@KN4CK3R

Quick summary of changes after last review:

1. Fixed access mode in push/delete API operations
2. Update to use package file stream instead of raw operations with object storage and sql database
3. Removed duplicated archive close operations
4. Moved repository automatic connection to [separate PR](https://github.com/go-gitea/gitea/pull/27318) in unified form for all package registries
5. Removed filename from composite key (correction related to unique index across multiple columns)
6. Refactoring in arch metadata module, shorter and more readable functions
7. Fixed error in pacman database creation function, when getting description property to write it in database
8. Corrections related to formatter and linter

d1nch8g commented:
@KN4CK3R 

More changes to unify formats with other package types:

1. Moved upload operation to arch service (would be easier to make [package upload UI](https://github.com/go-gitea/gitea/pull/26960))
2. Added metadata tests
3. Same import namings with other package types
4. Snake case for package metadata fields
5. Same error processing for response codes

d1nch8g commented:
@KN4CK3R 

Summary of changes since last review:

- Fixed access mode in upload/delete API operations
- Added a new endpoint to upload packages without signatures (pointing to the same method)
- Splitted metadata into version and file subtypes; file metadata contains only fields which can vary for different distros and arches.
- Metadata functions, fields, json tags and imports naming unification
- Error processing unification (same response codes with other package types)
- Additional metadata and integration tests for possible errors in metadata fields (checked with accordance with the PKGBUILD specification) and better integration tests for pacman database creation across different versions and architectures

This solution has many more tests and checks for rare conditions, and it is more consistent with other registries.Apologies if the prolonged implementation caused you any difficulties; the pull request should be ready for review at this point.

silverwind commented on docs/content/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,65 @@
+
+# Arch package registry
+
+Gitea has arch package registry, which can act as a fully working [arch linux mirror](https://wiki.archlinux.org/title/mirrors) and connected directly in `/etc/pacman.conf`. Gitea automatically creates pacman database for packages in user/organization space when new arch package is uploaded.
```
```suggestion
Gitea has a Arch Linux package registry, which can act as a fully working [Arch linux mirror](https://wiki.archlinux.org/title/mirrors) and connected directly in `/etc/pacman.conf`. Gitea automatically creates pacman database for packages in user/organization space when a new Arch package is uploaded.
```

silverwind commented on docs/content/usage/packages/arch.en-us.md:
````diff
@@ -0,0 +1,65 @@
+Server = https://{domain}/api/packages/{owner}/arch/{distribution}/{architecture}
+```
+
+Then, you can run pacman sync command (with -y flag to load connected database file), to install your package.
````
```suggestion
Then, you can run pacman sync command (with -y flag to load connected database file), to install your package:
```

silverwind commented on docs/content/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,65 @@
+
+## Upload packages
+
+When uploading package to gitea, you have to prepare package file with `.pkg.tar.zst` extension and it's `.pkg.tar.zst.sig` signature. You can use [curl](https://curl.se/) or any other HTTP client, gitea supports multiple [authentication schemes](https://docs.gitea.com/usage/authentication). Upload command will create 3 files: package, signature and desc file for pacman database (which will be cr ...
```
```suggestion
When uploading the package to gitea, you have to prepare package file with the `.pkg.tar.zst` extension and its `.pkg.tar.zst.sig` signature. You can use [curl](https://curl.se/) or any other HTTP client, Gitea supports multiple [authentication schemes](https://docs.gitea.com/usage/authentication). The upload command will create 3 files: package, signature and desc file for the pacman database (which will be created automatically on request).
```

silverwind commented on docs/content/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,65 @@
+
+When uploading package to gitea, you have to prepare package file with `.pkg.tar.zst` extension and it's `.pkg.tar.zst.sig` signature. You can use [curl](https://curl.se/) or any other HTTP client, gitea supports multiple [authentication schemes](https://docs.gitea.com/usage/authentication). Upload command will create 3 files: package, signature and desc file for pacman database (which will be cr ...
+
+Following command will upload arch package and related signature to gitea. Example with basic auth:
```
```suggestion
The following command will upload arch package and related signature to gitea with basic authentification:
```

silverwind commented on docs/content/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,65 @@
+
+## Delete packages
+
+Delete operation will remove specific package version, and all package files related to that version.
```
```suggestion
The `DELETE` method will remove specific package version, and all package files related to that version:
```

silverwind commented on docs/content/usage/packages/arch.en-us.md:
```diff
@@ -0,0 +1,65 @@
+
+## Clients
+
+Any `pacman` compatible package manager/AUR-helper can be used to install packages from gitea ([yay](https://github.com/Jguer/yay), [paru](https://github.com/Morganamilo/paru), [pikaur](https://github.com/actionless/pikaur), [aura](https://github.com/fosskers/aura)). Alternatively, you can try [pack](https://fmnx.su/core/pack) which supports full gitea API (install/push/remove). Also, any HTTP cl ...
```
```suggestion
Any `pacman` compatible package manager or AUR-helper can be used to install packages from gitea ([yay](https://github.com/Jguer/yay), [paru](https://github.com/Morganamilo/paru), [pikaur](https://github.com/actionless/pikaur), [aura](https://github.com/fosskers/aura)). Alternatively, you can try [pack](https://fmnx.su/core/pack) which supports full gitea API (install/push/remove). Also, any HTTP client can be used to execute get/push/remove operations ([curl](https://curl.se/), [postman](https://www.postman.com/), [thunder-client](https://www.thunderclient.com/)).
```

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -3338,6 +3338,10 @@ alpine.repository = Repository Info
 alpine.repository.architectures = Architectures
+arch.pacmanconf = Add server with related distribution and architecture to <code>/etc/pacman.conf</code>:
+arch.pacmansync = Sync package with pacman:
+arch.documentation = For more information on the arch mirrors, see <a target="_blank" rel="noopener noreferrer" href="%s">the documentation</a>.
```
Ideally we should not have HTML in the translation because later changes to the HTML are hard. One way of avoiding is to put the whole start and end tags token instead so this string becomes:

```ini
arch.documentation = For more information on the arch mirrors, see %sthe documentation%s.
```

And  the usage is:

```handlebars
{{.locale.Tr "packages.arch.documentation" (printf `<a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/usage/packages/arch/">`) (printf </a>) | Safe}}

silverwind commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,70 @@
+
+			{{if .PackageDescriptor.Metadata.Provides}}
+			<tr>
+				<td class="collapsing"><h5>Provides</h5></td>
```
I think we should likely translate these.

d1nch8g replied:
I was not sure about these; added translations just in case

silverwind commented on web_src/svg/gitea-arch.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" aria-hidden="true" class="octicon-arch__svg octicon-arch__gitea-arch svg octicon-arch" viewBox="0 0 128 128" width="16" height="16"><path d="M104.507 93.106c-15.886-29.291-38.862-73.3-40.98-82.79 0 0-4.614 11.632-12.844 29.554 2.996 4.643 10.373 9.645 13.073 12.855-9.321-2.941-14.997-8.793-14.997-8.793-9.577 20.631-23.104 47.827-38.444  ...
```
```suggestion
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="#1793d1" d="M256 72c-14 35-23 57-39 91 10 11 22 23 41 36-21-8-35-17-45-26-21 43-53 103-117 220 50-30 90-48 127-55-2-7-3-14-3-22v-1c1-33 18-58 38-56 20 1 36 29 35 62l-2 17c36 7 75 26 125 54l-27-50c-13-10-27-23-55-38 19 5 33 11 44 17-86-159-93-180-122-250z"/></svg>
```

Colored logo works better on dark theme.

silverwind reviewed (approved):
Looks good overall, can't really comment on go code though.

p0da commented:
Any update on this @d1nch8g?

d1nch8g commented:
@p0da

I am trying to add a package proxy and UI for different package files (should be easier to validate checksums).

ExplodingDragon commented:
Any new update on this @d1nch8g ?

d1nch8g commented:
@ExplodingDragon 

Sorry for the delayed response. Not entirely sure about the proper way to implement package signature verification.

Ideally, that should be visible for each package file, along with checksums and other file-related properties, in an alike way as it is done with commit signatures.

ExplodingDragon commented:
> @ExplodingDragon
> 
> Sorry for the delayed response. Not entirely sure about the proper way to implement package signature verification.
> 
> Ideally, that should be visible for each package file, along with checksums and other file-related properties, in an alike way as it is done with commit signatures.

This is not a problem if pacman can be skipped in the configuration file.

In fact,  rpm module doesn't do anything about signing rpm  packages file either. (https://github.com/go-gitea/gitea/pull/27069)
